libraries/AP_HAL_Linux/Thread.cpp

```diff
@@ -196,6 +196,24 @@ bool Thread::start(const char *name, int policy, int prio)
         AP_HAL::panic("Failed to create thread '%s': %s",
                       name, strerror(r));
     }
+
+    if (hal.util->get_affinity_core_num() > 0)
+    {
+        int current_core = hal.util->get_affinity_core_cur();
+    
+        cpu_set_t cpuset;
+        CPU_ZERO(&cpuset);
+        CPU_SET(current_core, &cpuset);
+    
+        if ((r = pthread_setaffinity_np(_ctx, sizeof(cpuset), &cpuset)) !=0) {
+            AP_HAL::panic("Failed to set affinity for thread '%s': %s",
+                          name, strerror(r));
+        }
+
+        
+        hal.util->set_affinity_core_cur(++current_core);
+    }
```

Comment:
the cpu affinity for threads should be inherited from the thread creating them... Since we are already doing it in the main thread, we shouldn't need this. Is this not the case for you?

Reply:
I am not sure I understood this point.
The issue is that the main thread will start by the system so I re allocated it to one of the cpu that are available for cpu_affinity.

Reply:
Yes. What I'm saying is that you only need to do it for the main thread, using sched_setaffinity(). Any thread created by pthread_create() inherits the cpuaffinity

Reply:
oic...but I create threads on different cpu...I think what you say is very valid on single dedicated cpu for all threads.
The proposed solution enable using multiple dedicated cpus.
I hope I understood what you said correctly.

Reply:
It doesn't matter how many cpus you are using. If you set it in the main thread all the others will inherit the CPU affinity

Reply:
> It doesn't matter how many cpus you are using. If you set it in the main thread all the others will inherit the CPU affinity

please bare with me.

I set RPI isolcpus=1,2 so now the any program will launch in cpu 0 & 3 only.

Now I use the command:

pi@raspberrypi:~ $ sudo /home/pi/arducopter -A udp:192.168.1.144:14550:bcast -B /dev/serial0 -C /dev/ardu --cpu-affinity 6

This command it instructs ardupilot to use cpu mask 0b0110 which means that core 1 & core 2 should be used.
What I expect from this command is to run all threads on cpu 1 & 2 and that what happens.

pi@raspberrypi:~ $ ps -mo pid,tid,%cpu,psr -C arducopter
PID TID %CPU PSR
1364 - 10.6 -
- 1364 4.6 1
- 1365 1.6 2
- 1366 0.3 1
- 1367 2.3 2
- 1368 0.0 1
- 1369 0.5 2
- 1370 1.5 1

Yes at the beginner I should move the main thread to cpu 1 or cpu 2 as it will be launched by OS on cpu 0 & 3 due to isolcpus values.

but I dont all threads to inherit the same cpui. I want to be able to put them on different cpus

Reply:
Please if you see that I am still cannot understand please let me know what I should do exactly. Thanks for your time and effort.

Reply:
when you set cpu affinity you are actually not moving the thread - you are just passing a set of CPUs the Linux scheduler is allowed to put them on. Alternative, and IMO simpler implementation: https://github.com/lucasdemarchi/ardupilot/commits/tip-cpuset

You should achieve the same result (although I didn't really test) by doing:

```
 sudo /home/pi/arducopter -A udp:192.168.1.144:14550:bcast -B /dev/serial0 -C /dev/ardu --cpu-affinity 1,2
```

Reply:
For RPI, another thing to remember is that there is the kernel spi thread. If you mix with isolcpus it means that  they will be running in CPUs that are not the same as the ones running ardupilot. Not saying it's bad, just saying that doing what we are doing we are not covering all the thread on the system.